### ISSUE TITLE
Issue 3884 "Undefined reference to `pthread_atfork'" not fixed

### ISSUE THREAD

**Issue by xetorixik** [openssl/openssl#5387]
First of all thanks @levitte for fixing issue 3884.
But. It just does not seem to be fixed. (at least not here)

After building an RPM/SRPM based on the latest Nginx.
The following error appears:

```
-Wl,-z,relro -Wl,-z,now -pie -ldl -lpthread -lpthread -lcrypt -lpcre openssl-1.1.1pre1/.openssl/lib/libssl.a openssl-1.1.1pre1/.openssl/lib/libcrypto.a -ldl -lz -Wl,-E
openssl-1.1.1pre1/.openssl/lib/libcrypto.a(threads_pthread.o): In function `fork_once_func':
threads_pthread.c:(.text+0x16): undefined reference to `pthread_atfork'
collect2: error: ld returned 1 exit status
make[1]: *** [objs/nginx] Error 1
make[1]: Leaving directory `/home/build/rpmbuild/BUILD/nginx-1.13.8'
make: *** [build] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.iEgI8e (%build)
```

There are 0 problems using the latest OpenSSL 1.1 production release, 
even the OpenSSL_1_1_0-stable branch (with the same fix for issue 3884) is working fine. 
It only applies to 1.1.1pre and up.

Going back to [issue 3884](https://github.com/openssl/openssl/issues/3884).
A few things in common.
User @meilihao got the same problem based on the same CentOS 7.4 system.
Contributor @richsalz suggests that it is in fact a Nginx problem(wrong config), why is this not yet reported? As the latest Nginx 1.13.8 is susceptible to this problem.

In short, the question: What is the problem?

Long live SRPM.
For anyone who wants to reproduce it, try the SRPM at: https://ufile.io/c6czp
And change openssl-1.1.1pre1-therealthing.tar.gz to openssl-1.1.1pre1.tar.gz before compiling (simple rename hack to compile the SRPM file).
Thanks.

**Comment by levitte** [comment]
1.1.1pre1 *is* a snapshot if the master branch, so I don't understand how one can be ok and the other not, unless there was a change on master since Tuesday that made things better for you...

**Comment by levitte** [comment]
I would like to know exactly how OpenSSL was configured, and the easiest would be if you attached your `configdata.pm` here.

**Comment by levitte** [comment]
... or gave us the output of `perl configdata.pm -d` if applicable.

**Comment by xetorixik** [comment]
> 1.1.1pre1 is a snapshot if the master branch, so I don't understand how one can be ok and the other not, unless there was a change on master since Tuesday that made things better for you...

As written above. It is about the master branch of OpenSSL 1.1 not OpenSSL 1.1.1.

> I would like to know exactly how OpenSSL was configured, and the easiest would be if you attached your configdata.pm here.

OpenSSL is compiled against Nginx only. With --with-openssl= and does not run on its own.
Therefore the SRPM as attachment so anyone can reproduce the steps, files, sripts etc it in his or hers VM.

**Comment by xetorixik** [comment]
Typo. It's the OpenSSL_1_1_0-stable branch not the master branch of OpenSSL 1.1.

**Comment by levitte** [comment]
You're doing things in the wrong order.  Basically, this line is superbly wrongly ordered:

```
-Wl,-z,relro -Wl,-z,now -pie -ldl -lpthread -lpthread -lcrypt -lpcre openssl-1.1.1pre1/.openssl/lib/libssl.a openssl-1.1.1pre1/.openssl/lib/libcrypto.a -ldl -lz -Wl,-E
```

See, libcrypto uses pthread features, so `-lpthread` should come *after* `openssl-1.1.1pre1/.openssl/lib/libcrypto.a`

So looking more closely at `nginx-1.13.8/auto/lib/openssl/conf`, [I notice that you obviously know you need to add `-ldl`](https://github.com/nginx/nginx/blob/master/auto/lib/openssl/conf#L43).  So very simple solution, you need to do something similar with `-lpthread`.

Now guys, *this isn't an OpenSSL issue at all*, and I just spent time doing someone else's homework for you.  Please turn to [the nginx project](https://github.com/nginx/nginx) with this sort of problem unless you can clearly see that it's the *OpenSSL build* that fails.

Closing this.
